### PR TITLE
feat(appconfig): add grantReadConfig method to Environment Construct

### LINKED ISSUE #28585: (appconfig-alpha): grant read Configuration
### Describe the feature

Now `appconfig-alpha` has no grant method. If `grantReadConfig()` is implemented, it will be useful.

### Use Case

Using AppConfig from application codes.

### Proposed Solution

```ts
environment.grantReadConfig(lambdaFunction);
```

### Other Information

Needed policy is following:
```ts
new iam.PolicyStatement({
  actions: [
    "appconfig:GetLatestConfiguration",
    "appconfig:StartConfigurationSession",
  ],
  resources: [`${appConfigEnvironment.environmentArn}/*`],
})
```

see, https://docs.aws.amazon.com/appconfig/latest/userguide/about-data-plane.html

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.114.1

### Environment details (OS name and version, etc.)

Mac OS, Apple silicon, Sonoma 14.2.1

### PR DESCRIPTION
**Pull request by mazyu36**
### Issue # (if applicable)

Closes #28585

### Reason for this change
To simplify policy configuration for AppConfig Environments.

### Description of changes
add grantReadConfig method to Environment Construct

### Description of how you validated changes
Add unit test and integ tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 1]
```diff
@@ -543,4 +559,19 @@ export interface IEnvironment extends IResource {
...
   addExtension(extension: IExtension): void;
+
+  /**
+   * Grant the given identity the specified actions
+   * @param grantee the identity to be granted the actions
+   * @param actions the data-access actions
+   */
+  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+
+  /** TODO
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:570]: I guess this TODO can be removed?

Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/lib/environment.ts]: @daschaa 
Oops...sorry, I removed it.

File: packages/aws-cdk-lib/aws-appconfig/README.md [hunk 2]
```diff
@@ -78,6 +78,24 @@ new appconfig.Environment(this, 'MyEnvironment', {
 Environment monitors also support L1 `CfnEnvironment.MonitorsProperty` constructs through the `fromCfnMonitorsProperty` method.
 However, this is not the recommended approach for CloudWatch alarms because a role will not be auto-generated if not provided.
 
+See [About the AWS AppConfig data plane service](https://docs.aws.amazon.com/appconfig/latest/userguide/about-data-plane.html) more information.
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/README.md:81]: *for more information

File: packages/aws-cdk-lib/aws-appconfig/README.md [hunk 3]
```diff
@@ -78,6 +78,24 @@ new appconfig.Environment(this, 'MyEnvironment', {
...
+
+### Permissions
+
+You can grant read configuration permission on an Environment to any Principal as follows:
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/README.md:85]: This sounds a bit raw to me. Maybe: You can grant permissions to read a configuration to an Environment with the `grantReadConfig` method as follows:

What do you think?

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 4]
```diff
@@ -543,4 +559,19 @@ export interface IEnvironment extends IResource {
...
+  /**
+   * Grant the given identity the specified actions
+   * @param grantee the identity to be granted the actions
+   * @param actions the data-access actions
+   */
+  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+
+  /**
+   * Grants read configuration permissions for this environment
+   * to an IAM principal (Role/Group/User).
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:572]: Is the `Grantable` interface === IAM principal? Or is it even broader? 
Are other docstrings for grant methods also describing this in this way?

File: packages/aws-cdk-lib/aws-appconfig/README.md (hunk 2 above)
Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/README.md]: @daschaa 
Thank you for your review.
I added `for`.

File: packages/aws-cdk-lib/aws-appconfig/README.md (hunk 3 above)
Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/README.md]: @daschaa 
I think your suggestion is better.
I edited it.

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts (hunk 4 above)
Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/lib/environment.ts]: @daschaa 
From my understanding, the Grantable interface === IAM principal.
I have revised the explanation referring to the documentation of the grant method of the DynamoDB Table v2 Construct.

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_dynamodb.TableV2.html#grantwbrreadwbrwritewbrdatagrantee

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 5]
```diff
@@ -561,15 +561,17 @@ export interface IEnvironment extends IResource {
   addExtension(extension: IExtension): void;
 
   /**
-   * Grant the given identity the specified actions
-   * @param grantee the identity to be granted the actions
-   * @param actions the data-access actions
+   * Adds an IAM policy statement associated with this environment to an IAM principal's policy.
+   *
+   * @param grantee the principal (no-op if undefined)
+   * @param actions the set of actions to allow (i.e., 'appconfig:GetLatestConfiguration', 'appconfig:StartConfigurationSession', etc.)
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:567]: I think the comma after i.e. can be removed, right? 🤔

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 6]
```diff
@@ -561,15 +561,17 @@ export interface IEnvironment extends IResource {
...
    */
   grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
 
   /**
-   * Grants read configuration permissions for this environment
-   * to an IAM principal (Role/Group/User).
+   * Permits an IAM principal a configuration read operations on this environment.
```
Comment (daschaa) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:572]: I think how you wrote it before `Grants ...` was already good. My comment was rather a question if the term `IAM principal` is correct here? I'm not sure if an IAM principal is the same scope like an Grantable. I was expecting a comment like `Grants a resource read permission on the configurations of this environment`
What do you think? I would be happy to hear your opinion here :)

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts (hunk 4 above)
Comment (daschaa) [reply packages/aws-cdk-lib/aws-appconfig/lib/environment.ts]: @mazyu36 Awesome! Thank you very much for the explanation!

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts (hunk 5 above)
Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:567]: @daschaa 
I'm adding a comma (,) to align with the descriptions for DynamoDB.

https://github.com/aws/aws-cdk/blob/4b6dc8c650822bcd0231c8890bd94a934a0cd34d/packages/aws-cdk-lib/aws-dynamodb/lib/table-v2-base.ts#L69

**Comment by daschaa** [comment]
@mazyu36 Thank you very much for the pull requests and sorry for the back and forth with the docstrings.

**Review by daschaa — APPROVED** [review]

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 7]
```diff
@@ -116,6 +116,22 @@ abstract class EnvironmentBase extends Resource implements IEnvironment, IExtens
   public addExtension(extension: IExtension) {
     this.extensible.addExtension(extension);
   }
+
+  public grant(grantee: iam.IGrantable, ...actions: string[]) {
+    return iam.Grant.addToPrincipal({
+      grantee,
+      actions,
+      resourceArns: [`${this.environmentArn}/*`],
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:124]: I'm not too familiar with appconfig usage. Any document supporting that this is the correct arns and least privilege?

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts [hunk 8]
```diff
@@ -543,4 +559,21 @@ export interface IEnvironment extends IResource {
...
+  /**
+   * Adds an IAM policy statement associated with this environment to an IAM principal's policy.
+   *
+   * @param grantee the principal (no-op if undefined)
+   * @param actions the set of actions to allow (i.e., 'appconfig:GetLatestConfiguration', 'appconfig:StartConfigurationSession', etc.)
+   */
+  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+
+  /**
+   * Permits an IAM principal a configuration read operations on this environment.
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-appconfig/lib/environment.ts:572]:
```suggestion
   * Permits an IAM principal to perform read operations on this environment's configuration.
```

File: packages/aws-cdk-lib/aws-appconfig/README.md [hunk 9]
```diff
@@ -3,13 +3,13 @@
 
 This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.
 
-For a high level overview of what AWS AppConfig is and how it works, please take a look here:
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-appconfig/README.md:6]: Nice, thanks for removing the whitespaces.

**Review by GavinZZ — COMMENTED** [review]
LGTM in general, just one question regarding permission.

File: packages/aws-cdk-lib/aws-appconfig/lib/environment.ts (hunk 7 above)
Comment (mazyu36) [reply packages/aws-cdk-lib/aws-appconfig/lib/environment.ts]: Thank you for your comment.

Upon reviewing it again, I found that the scope of the `resourcesArn` for `grantReadConfig` was too broad, so I made a modification (added `/configuration/`).

In AppConfig, a Configuration Profile represents a parameter (e.g., a Feature Flag), and an Environment represents a logical group of deployed Configuration Profiles.  ([docs](https://docs.aws.amazon.com/appconfig/latest/userguide/what-is-appconfig.html#learn-more-appconfig-how-it-works))

The ARN of a Configuration belonging to an Environment takes the following format.  ([docs](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsappconfig.html#awsappconfig-resources-for-iam-policies))

> arn:${Partition}:appconfig:${Region}:${Account}:application/${ApplicationId}/environment/${EnvironmentId}/configuration/${ConfigurationProfileId}

For this issue, I wanted to grant read permission for the parameters (Configurations) belonging to a logical group (Environment), so I needed to use a wildcard for `/configuration/ `and beyond, as I realized.

Previously, the wildcard was applied from `/${EnvironmentId}/` onward, which was too broad in scope.

Regarding the `grant` method, I changed it to limit the resources to the Environment only.

**Comment by mazyu36** [comment]
@GavinZZ 
Thank you for your review!
I've addressed all your comments.

**Review by GavinZZ — APPROVED** [review]
LGTM, thanks for contributing!

**Pull request merged**